File: lambda/mcp-server/s3_client.py
"""S3 client for accessing Krisp meeting transcripts."""

import os
import json
import re
from datetime import datetime, timedelta
from typing import List, Optional, Dict, Any

import boto3
from botocore.config import Config

BUCKET_NAME = os.environ.get("KRISP_S3_BUCKET", "")  # Required: set via environment variable
AWS_REGION = os.environ.get("AWS_REGION", "us-east-1")

# Key format regex: YYYYMMDD_HHMMSS_title_meetingId.json
KEY_PATTERN = re.compile(r"^(\d{8})_(\d{6})_(.+)_([^_]+)\.json$")


class S3TranscriptClient:
    """Client for S3 transcript operations."""

    def __init__(self):
        self.s3 = boto3.client(
            "s3",
            region_name=AWS_REGION,
            config=Config(retries={"max_attempts": 3, "mode": "adaptive"}),
        )
        self.bucket = BUCKET_NAME

    def list_transcripts(
        self,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        limit: int = 20,
    ) -> List[Dict[str, Any]]:
        """List transcripts within date range."""
        if end_date is None:
            end_date = datetime.utcnow()
        if start_date is None:
            start_date = end_date - timedelta(days=30)

        # Generate prefixes for efficient S3 listing
        prefixes = self._generate_date_prefixes(start_date, end_date)

        all_objects = []
        for prefix in prefixes:
            paginator = self.s3.get_paginator("list_objects_v2")
            for page in paginator.paginate(Bucket=self.bucket, Prefix=prefix):
                for obj in page.get("Contents", []):
                    if obj["Key"].endswith(".json"):
                        metadata = self._parse_key_metadata(obj["Key"], obj)
                        if metadata and start_date <= metadata["date"] <= end_date:
                            all_objects.append(metadata)

        # Sort by date descending, apply limit
        all_objects.sort(key=lambda x: x["date"], reverse=True)
        return all_objects[:limit]

    def get_transcript(self, key: str) -> Dict[str, Any]:
        """Fetch full transcript content by S3 key."""
        response = self.s3.get_object(Bucket=self.bucket, Key=key)
        content = json.loads(response["Body"].read().decode("utf-8"))
        return content

    def get_transcripts(self, keys: List[str]) -> List[Dict[str, Any]]:
        """Fetch multiple transcripts by S3 keys."""
        results = []
        for key in keys:
            try:
                content = self.get_transcript(key)
                raw = content.get("raw_payload", {})
                results.append({
                    "key": key,
                    "title": raw.get("title", "Untitled"),
                    "summary": raw.get("summary", ""),
                    "notes": raw.get("notes", ""),
                    "transcript": raw.get("transcript", ""),
                    "action_items": raw.get("action_items", []),
                    "speakers": raw.get("speakers", []),
                    "received_at": content.get("received_at", ""),
                    "event_type": content.get("event_type", ""),
                    "error": None,
                })
            except Exception as e:
                results.append({"key": key, "error": str(e)})
        return results

    def search(
        self,
        query: str,
        speaker: Optional[str] = None,
        limit: int = 10,
    ) -> List[Dict[str, Any]]:
        """Search transcripts by content and/or speaker."""
        # Search across last 90 days
        recent = self.list_transcripts(
            start_date=datetime.utcnow() - timedelta(days=90),
            end_date=datetime.utcnow(),
            limit=200,  # Search pool
        )

        results = []
        query_lower = query.lower()

        for meta in recent:
            try:
                content = self.get_transcript(meta["key"])
                raw = content.get("raw_payload", {})

                # Check speaker filter
                if speaker:
                    speakers = raw.get("speakers", [])
                    speaker_names = [s.get("name", "").lower() for s in speakers]
                    if not any(speaker.lower() in name for name in speaker_names):
                        continue

                # Search in relevant fields
                searchable = " ".join([
                    raw.get("transcript", ""),
                    raw.get("summary", ""),
                    raw.get("notes", ""),
                    raw.get("title", ""),
                ]).lower()

                if query_lower in searchable:
                    snippet = self._extract_snippet(searchable, query_lower)
                    results.append({
                        **meta,
                        "snippet": snippet,
                        "summary": raw.get("summary", "")[:300],
                        "speakers": [s.get("name", "") for s in raw.get("speakers", [])],
                    })

                if len(results) >= limit:
                    break
            except Exception:
                continue  # Skip failed fetches

        return results

    def _parse_key_metadata(self, key: str, obj: dict) -> Optional[Dict[str, Any]]:
        """Parse metadata from S3 key format."""
        parts = key.split("/")
        if len(parts) < 2:
            return None

        filename = parts[-1]
        match = KEY_PATTERN.match(filename)

        if match:
            date_str, time_str, title, meeting_id = match.groups()
            date = datetime.strptime(f"{date_str}_{time_str}", "%Y%m%d_%H%M%S")
            title = title.replace("_", " ")
        else:
            date = obj.get("LastModified", datetime.utcnow())
            if hasattr(date, "replace"):
                date = date.replace(tzinfo=None)
            title = filename.replace(".json", "")
            meeting_id = ""

        return {
            "key": key,
            "title": title,
            "meeting_id": meeting_id,
            "date": date,
            "date_str": date.strftime("%Y-%m-%d %H:%M"),
            "size": obj.get("Size", 0),
        }

    def _generate_date_prefixes(
        self, start: datetime, end: datetime
    ) -> List[str]:
        """Generate S3 prefixes for date range."""
        prefixes = set()
        current = start
        while current <= end:
            prefixes.add(f"meetings/{current.strftime('%Y/%m/')}")
            # Move to next month
            if current.month == 12:
                current = current.replace(year=current.year + 1, month=1, day=1)
            else:
                current = current.replace(month=current.month + 1, day=1)
        return sorted(prefixes)

    def _extract_snippet(self, text: str, query: str, context: int = 100) -> str:
        """Extract text snippet around search match."""
        idx = text.find(query)
        if idx == -1:
            return text[:200] + "..." if len(text) > 200 else text

        start = max(0, idx - context)
        end = min(len(text), idx + len(query) + context)
        snippet = text[start:end]

        if start > 0:
            snippet = "..." + snippet
        if end < len(text):
            snippet = snippet + "..."

        return snippet

File: lambda/mcp-server-ts/src/s3-client.ts
/**
 * S3 client for accessing Krisp meeting transcripts.
 */

import { S3Client, ListObjectsV2Command, GetObjectCommand } from '@aws-sdk/client-s3';

const BUCKET_NAME = process.env.KRISP_S3_BUCKET || '';  // Required: set via environment variable
const AWS_REGION = process.env.AWS_REGION || 'us-east-1';

// Key format regex: YYYYMMDD_HHMMSS_title_meetingId.json
const KEY_PATTERN = /^(\d{8})_(\d{6})_(.+)_([^_]+)\.json$/;

interface TranscriptMetadata {
  key: string;
  title: string;
  meetingId: string;
  date: Date;
  dateStr: string;
  size: number;
}

interface TranscriptContent {
  key: string;
  title: string;
  summary: string;
  notes: string;
  transcript: string;
  actionItems: string[];
  speakers: string[];
  receivedAt: string;
  eventType: string;
  error: string | null;
}

interface SearchResult extends TranscriptMetadata {
  snippet: string;
  summary: string;
  speakers: string[];
}

export class S3TranscriptClient {
  private s3: S3Client;
  private bucket: string;

  constructor() {
    this.s3 = new S3Client({ region: AWS_REGION });
    this.bucket = BUCKET_NAME;
  }

  async listTranscripts(
    startDate?: Date,
    endDate?: Date,
    limit: number = 20
  ): Promise<TranscriptMetadata[]> {
    const end = endDate || new Date();
    const start = startDate || new Date(end.getTime() - 30 * 24 * 60 * 60 * 1000);

    const prefixes = this.generateDatePrefixes(start, end);
    const allObjects: TranscriptMetadata[] = [];

    for (const prefix of prefixes) {
      let continuationToken: string | undefined;

      do {
        const command = new ListObjectsV2Command({
          Bucket: this.bucket,
          Prefix: prefix,
          ContinuationToken: continuationToken,
        });

        const response = await this.s3.send(command);

        for (const obj of response.Contents || []) {
          if (obj.Key?.endsWith('.json')) {
            const metadata = this.parseKeyMetadata(obj.Key, obj);
            if (metadata && metadata.date >= start && metadata.date <= end) {
              allObjects.push(metadata);
            }
          }
        }

        continuationToken = response.NextContinuationToken;
      } while (continuationToken);
    }

    // Sort by date descending, apply limit
    allObjects.sort((a, b) => b.date.getTime() - a.date.getTime());
    return allObjects.slice(0, limit);
  }

  async getTranscript(key: string): Promise<unknown> {
    const command = new GetObjectCommand({
      Bucket: this.bucket,
      Key: key,
    });

    const response = await this.s3.send(command);
    const bodyString = await response.Body?.transformToString();
    return JSON.parse(bodyString || '{}');
  }

  async getTranscripts(keys: string[]): Promise<TranscriptContent[]> {
    const results: TranscriptContent[] = [];

    for (const key of keys) {
      try {
        const content = await this.getTranscript(key) as Record<string, unknown>;
        const rawPayload = (content.raw_payload || {}) as Record<string, unknown>;
        const data = (rawPayload.data || {}) as Record<string, unknown>;
        const meeting = (data.meeting || {}) as Record<string, unknown>;
        const speakers = (meeting.speakers || []) as Array<{ first_name?: string; last_name?: string; index?: number }>;

        results.push({
          key,
          title: (meeting.title as string) || 'Untitled',
          summary: (data.raw_meeting as string) || '',
          notes: '',
          transcript: (data.raw_content as string) || '',
          actionItems: [],
          speakers: speakers.map(s => s.first_name ? `${s.first_name} ${s.last_name || ''}`.trim() : `Speaker ${s.index}`),
          receivedAt: (content.received_at as string) || '',
          eventType: (content.event_type as string) || '',
          error: null,
        });
      } catch (e) {
        results.push({
          key,
          title: '',
          summary: '',
          notes: '',
          transcript: '',
          actionItems: [],
          speakers: [],
          receivedAt: '',
          eventType: '',
          error: String(e),
        });
      }
    }

    return results;
  }

  async search(
    query: string,
    speaker?: string,
    limit: number = 10
  ): Promise<SearchResult[]> {
    // Search across last 90 days
    const recent = await this.listTranscripts(
      new Date(Date.now() - 90 * 24 * 60 * 60 * 1000),
      new Date(),
      200
    );

    const results: SearchResult[] = [];
    const queryLower = query.toLowerCase();

    for (const meta of recent) {
      try {
        const content = await this.getTranscript(meta.key) as Record<string, unknown>;
        const rawPayload = (content.raw_payload || {}) as Record<string, unknown>;
        const data = (rawPayload.data || {}) as Record<string, unknown>;
        const meeting = (data.meeting || {}) as Record<string, unknown>;
        const speakersList = (meeting.speakers || []) as Array<{ first_name?: string; last_name?: string; index?: number }>;

        // Check speaker filter
        if (speaker) {
          const speakerNames = speakersList.map(s =>
            s.first_name ? `${s.first_name} ${s.last_name || ''}`.toLowerCase() : ''
          );
          if (!speakerNames.some(name => name.includes(speaker.toLowerCase()))) {
            continue;
          }
        }

        // Search in relevant fields
        const searchable = [
          data.raw_content || '',
          data.raw_meeting || '',
          meeting.title || '',
        ].join(' ').toLowerCase();

        if (searchable.includes(queryLower)) {
          const snippet = this.extractSnippet(searchable, queryLower);
          results.push({
            ...meta,
            snippet,
            summary: ((data.raw_meeting as string) || '').slice(0, 300),
            speakers: speakersList.map(s => s.first_name ? `${s.first_name} ${s.last_name || ''}`.trim() : `Speaker ${s.index}`),
          });
        }

        if (results.length >= limit) {
          break;
        }
      } catch {
        continue;
      }
    }

    return results;
  }

  private parseKeyMetadata(
    key: string,
    obj: { LastModified?: Date; Size?: number }
  ): TranscriptMetadata | null {
    const parts = key.split('/');
    if (parts.length < 2) return null;

    const filename = parts[parts.length - 1];
    const match = filename.match(KEY_PATTERN);

    let date: Date;
    let title: string;
    let meetingId: string;

    if (match) {
      const [, dateStr, timeStr, rawTitle, id] = match;
      date = new Date(
        parseInt(dateStr.slice(0, 4)),
        parseInt(dateStr.slice(4, 6)) - 1,
        parseInt(dateStr.slice(6, 8)),
        parseInt(timeStr.slice(0, 2)),
        parseInt(timeStr.slice(2, 4)),
        parseInt(timeStr.slice(4, 6))
      );
      title = rawTitle.replace(/_/g, ' ');
      meetingId = id;
    } else {
      date = obj.LastModified || new Date();
      title = filename.replace('.json', '');
      meetingId = '';
    }

    return {
      key,
      title,
      meetingId,
      date,
      dateStr: date.toISOString().slice(0, 16).replace('T', ' '),
      size: obj.Size || 0,
    };
  }

  private generateDatePrefixes(start: Date, end: Date): string[] {
    const prefixes = new Set<string>();
    const current = new Date(start);

    while (current <= end) {
      const year = current.getFullYear();
      const month = String(current.getMonth() + 1).padStart(2, '0');
      prefixes.add(`meetings/${year}/${month}/`);

      // Move to next month
      current.setMonth(current.getMonth() + 1);
      current.setDate(1);
    }

    return Array.from(prefixes).sort();
  }

  private extractSnippet(text: string, query: string, context: number = 100): string {
    const idx = text.indexOf(query);
    if (idx === -1) {
      return text.length > 200 ? text.slice(0, 200) + '...' : text;
    }

    const start = Math.max(0, idx - context);
    const end = Math.min(text.length, idx + query.length + context);
    let snippet = text.slice(start, end);

    if (start > 0) snippet = '...' + snippet;
    if (end < text.length) snippet = snippet + '...';

    return snippet;
  }
}
